staging/src/k8s.io/api/storage/v1beta1/types.go

```diff
@@ -166,7 +166,12 @@ type VolumeAttachmentSource struct {
 	// +optional
 	PersistentVolumeName *string `json:"persistentVolumeName,omitempty" protobuf:"bytes,1,opt,name=persistentVolumeName"`
 
-	// Placeholder for *VolumeSource to accommodate inline volumes in pods.
+	// Copy of VolumeSource to accommodate VolumeSource from a pod
+	// to support shimming of in-tree inline volumes to a CSI backend.
+	// This field is alpha-level and is only honored by servers that
+	// enable the CSIMigration feature.
+	// +optional
+	InlineVolumeSource *v1.VolumeSource `json:"inlineVolumeSource,omitempty" protobuf:"bytes,2,opt,name=inlineVolumeSource"`
```

Comment:
Could this plausibly be a CSI spec inside here?

Reply:
Not exactly sure which question you're asking so I will answer it the two ways I understood:
We are not adding the CSI Spec inside of the `VolumeSource` struct so we are not leaking CSI VolumeSource into this field.
The `CSIVolumeSource` (https://github.com/kubernetes/kubernetes/pull/74086/commits/5b64d68c5141de77acbda0de94b0a138fe6ce44f#diff-4143fe2cdf42f1748648a9b80aa2ba07R1610) is stripped down to only support `mount/unmount` fields so is not robust enough for all the information the external-attacher needs for shimmed volumes. Therefore we will just send the un-translated in-tree volume source to the external-attacher and translate it on the external-attacher side.

Hopefully that answered your question.

Reply:
could it be a `*CSIPersistentVolumeSource`?

Reply:
I guess I was just saying it's sort of recursive.  Not really a problem.

Reply:
We can explicitly disallow `CSIVolumeSource` in here.  `CSIPersistentVolumeSource` is only part of `PersistentVolumeSource`

Reply:
> `CSIPersistentVolumeSource` is only part of `PersistentVolumeSource`

I know, but the information produced by the inline volume -> csi translator is all present in the CSIPersistentVolumeSource type: namespaced secret references, full volume options, etc

I was asking if it would make sense to use `*CSIPersistentVolumeSource` in this field *instead of* `*VolumeSource`

Reply:
Oh that is very intriguing. Will let @davidz627 evaluate.

Reply:
Really interesting idea, and I see how this would solve: https://github.com/kubernetes/kubernetes/pull/74629#discussion_r260589501 as well...

Then we can restrict the allowed volume sources based on feature flags in validation to resolve https://github.com/kubernetes/kubernetes/pull/74629#discussion_r260589421 (as well as always restrict the `CSIPersistentVolumeSource`). I like this idea. Will implement

Reply:
misread comment. I see `CSIPersistentVolumeSource` now. I believe that was the original design but we lost our way in the middle there.

It makes the most sense because it contains all the namespace information. As well as being the actual end artifact required for the external-attacher to do its thing. It makes more sense to do the translation of the in-line volume on the Kubernetes side where we have more of the Pod information than trying to do it on the external attacher side. This also simplifies the API validation.